src/core/slice.c

```diff
@@ -413,6 +413,26 @@ static bool slice_can_freeze(Unit *s) {
         return slice_freezer_action_supported_by_children(s);
 }
 
+static void slice_notify_cgroup_empty_event(Unit *u) {
+        Slice *s = SLICE(u);
+
+        assert(u);
+
+        log_unit_debug(u, "Control group is empty.");
+
+        switch (s->state) {
+
+        /* If the cgroup empty notification comes when the unit is not active, we must have failed to clean
+         * up the cgroup earlier and should do it now. */
+        case SLICE_DEAD:
+                unit_prune_cgroup(u);
+                break;
+
+        default:
+                ;
+        }
+}
```

Comment:
hmm, i scope and service units have this already, and i figure all units with cgroups should really have this.

Hence I think we should really make this generic. i.e. move this to on_cgroup_empty_event():

```c
if (UNIT_IS_INACTIVE_OR_DEAD(unit_status(u)) 
        unit_prune_cgroup(u);
```

and then don't even pass control to the per-unit-type handlers for these states, and remove the relevant handling from service/scope/slice code for all states that map to UNIT_DEAD/UNIT_FAILED.

Reply:
also, this really should be a separate commit in this PR.